gtsam/hybrid/HybridGaussianFactorGraph.cpp

```diff
@@ -469,33 +439,25 @@ AlgebraicDecisionTree<Key> HybridGaussianFactorGraph::error(
   AlgebraicDecisionTree<Key> error_tree(0.0);
 
   // Iterate over each factor.
-  for (size_t idx = 0; idx < size(); idx++) {
+  for (auto &f : factors_) {
     // TODO(dellaert): just use a virtual method defined in HybridFactor.
     AlgebraicDecisionTree<Key> factor_error;
 
-    if (factors_.at(idx)->isHybrid()) {
-      // If factor is hybrid, select based on assignment.
-      GaussianMixtureFactor::shared_ptr gaussianMixture =
-          boost::static_pointer_cast<GaussianMixtureFactor>(factors_.at(idx));
+    if (auto gaussianMixture = dynamic_pointer_cast<GaussianMixtureFactor>(f)) {
```

Comment:
This makes me sad. `f->isHybrid()` is much cleaner and self-documenting. This is just hard to read and more intimidating (e.g. if I was a new contributor).

Reply:
It should be handled with virtual methods. Next PR